numba/core/untyped_passes.py

```diff
@@ -1680,6 +1681,135 @@ def run_pass(self, state):
+                if expr:
+                    if isinstance(expr, ir.Expr) and expr.op == 'getiter':
+                        # expr := getiter(value=Var)
+                        if self.is_arg_constsized(state, expr.value) or \
+                           self.match_iter(state, expr):
+                            dead_loops.append(loop)
+        return dead_loops
+
+    def remove_dead_loops(self, state, dead_loops):
+        """
+        For each dead loop, change the branch to a direct jump to the exit
+        block. This function also removes the `pair_first` instruction, if
+        exists, since DCE won't remove it.
+        """
+        for loop in dead_loops:
+            # change branch into a direct jump to exit
+            header = state.func_ir.blocks[loop.header]
+            term = header.terminator
+            header.insert_before_terminator(ir.Jump(term.falsebr, term.loc))
+            header.remove(term)
+
+            # iterate over header and remove "pair_first", since
+            # dead_code_elimination won't remove it
+            for inst in header.find_insts(ir.Assign):
+                expr = inst.value
+                if isinstance(expr, ir.Expr) and expr.op == 'pair_first':
+                    header.remove(inst)
+                    if guard(get_definition, state.func_ir, inst.target.name):
+                        state.func_ir.remove_definition(inst.target.name)
+
+        dead_branch_prune(state.func_ir, state.args)
+
+        for lhs in self.dead_vars:
+            if lhs is None:
+                continue
+            for block in state.func_ir.blocks.values():
+                # find the assignment associated with lhs
+                assign = block.find_variable_assignment(lhs.name)
+                if assign:
+                    block.remove(assign)
+                    if guard(get_definition, state.func_ir, assign.target.name):
+                        state.func_ir.remove_definition(assign.target.name)
+
+    def run_pass(self, state):
+        # run as subpipeline
+        from numba.core.compiler_machinery import PassManager
+        from numba.core.typed_passes import PartialTypeInference
+        pm = PassManager("dead_loop_elimination_subpipeline")
+        if not state.flags.no_rewrites:
+            pm.add_pass(RewriteSemanticConstants, "Semantic const")
+            pm.add_pass(DeadBranchPrune, "Dead branch prune")
+            pm.add_pass(GenericRewrites, "Generic Rewrites")
+        pm.add_pass(PartialTypeInference, "performs partial type inference")
+        pm.finalize()
+        pm.run(state)
+
+        self.dead_vars = []
+        dead_loops = self.find_dead_loops(state)
+        if dead_loops:
+            self.remove_dead_loops(state, dead_loops)
+            state.func_ir.blocks = simplify_CFG(state.func_ir.blocks)
+            dead_code_elimination(state.func_ir)
+
+            # reset type inference now we are done with the partial results
+            state.typemap = None
```

Comment:
```suggestion
            state.typemap = {}
```

Reply:
In other places, `state.typemap` is assigned to `None`. Maybe as an effort to standardize, one should change None values to `{}`.